design/go-api.md

```diff
@@ -0,0 +1,72 @@
+# Replace the current typescript kurl.sh api for creating installers with a go api using gin
+ 
+This proposal outlines the benefits and challenges of replacing the current typescript api with a more simple golang api.
+
+## Goals
+
+- Significantly reduce code complexety around making changes to the installer type
+- Provide type safety and validation for POSTed installers
+
+## Non Goals
+
+- Create an API for kurlkinds dsecriptions. This probably best suited in the kurl.sh site repo as it is not core functionality of the api.
+
+## Background
+
+Previously, kURL has used the backend typescript api to take a yaml config and make an install script with options and bash variables set for the install.
+With the recent changes to kURL such as the addition of the CRD, the current install script uses go binaries to parse and traverse the yaml tree to set options.
+
+We can greatly reduce code complexity from the no longer used parts of the typescript api and gain added type safety by moving this api to Go.
+
+## High-Level Design
+
+Endpoints
+
+postInstaller POST /installer
+putInstaller PUT /installer/{name}
+getInstallScript GET /{installerID}
+getJoinScript GET /{installerID}/join.sh
+getUpgradeScript GET /{installerID}/upgrade.sh
+getInstallerYaml GET /installer/{installerID}
+validateInstaller POST /installer/validate
+getBundle GET/bundle/{installerID}
+
+## Detailed Design
+
+postInstaller Endpoint:
+The function will take a yaml Installer and attempt to unmarshall it into an Installer CRD.
+If the unmarshalling fails, an error will be retuned. If the marshalling succeeds the version of an addon will be checked against existing versions.
+If the version is not found an error will be returned.
+Any version latest will be converted to the most recent version.
+
+putInstaller Endpoint
+This endpoint will take yaml installer, do the validation as done inpostInstaller, and update that installer in the db.
+This endpoint will also handle JWT auth in order to validate the creater of the installer. 
+
+getInstallScript, getJoinScript, getUpgradeScript Endpoints:
+These will check the db for a specific installerID, and inject it into the script as INSTALLER_SPEC_YAML bash strings
+
+validateInstaller Endpoint:
+This endpoint will take a yaml file and do the steps similar to postInstaller Endpoint and return a response indicating if it is invalid or not.
+
+getBundle Endpoint:
+This is already served by a golang endpoint
```

Comment:
Currently it calls the typescript API to get the rendered scripts and the list of packages to push into the bundle.

Reply:
noted